lib/Filters/GitDiffFilter.cpp

```diff
@@ -0,0 +1,97 @@
+#include "mull/Filters/GitDiffFilter.h"
+
+#include "mull/Diagnostics/Diagnostics.h"
+#include "mull/Toolchain/Runner.h"
+
+#include "llvm/IR/InstIterator.h"
+#include <llvm/IR/DebugInfoMetadata.h>
+#include <llvm/IR/DebugLoc.h>
+#include <llvm/IR/Function.h>
+#include <llvm/Support/FileSystem.h>
+
+#include <iostream>
+#include <regex>
+#include <sstream>
+#include <unistd.h>
+
+using namespace mull;
+
+static std::string getCwd() {
+  char cwd_buffer[8096];
+  if (getcwd(cwd_buffer, sizeof(cwd_buffer)) == NULL) {
+    return "";
+  }
+  return std::string(cwd_buffer);
+}
+
+GitDiffFilter::GitDiffFilter(Diagnostics &diagnostics)
+    : diagnostics(diagnostics), gitDiffInfoStorage() {}
+
+std::string GitDiffFilter::name() {
+  return "Git Diff";
+}
+
+void GitDiffFilter::activateAgainstGitBranch(const std::string &gitDiffBranch) {
+  /// Assuming that Mull is run from the project's directory.
+  const std::string gitRootPath = getCwd();
+  mull::GitDiffReader gitDiffReader(diagnostics, gitRootPath);
+
+  mull::GitDiffInfo gitDiffInfo = gitDiffReader.readGitDiff(gitDiffBranch);
+
+  gitDiffInfoStorage = gitDiffInfo;
+}
+
+bool GitDiffFilter::shouldSkip(llvm::Function *function) {
+  auto debugInfo = llvm::dyn_cast<llvm::DISubprogram>(function->getMetadata(0));
+
+  std::string unitDirectory;
+  std::string unitFilePath;
+  if (llvm::DICompileUnit *unit = debugInfo->getUnit()) {
+    unitDirectory = unit->getDirectory();
+    unitFilePath = unit->getFilename();
+  }
+  assert(unitFilePath[0] == '/');
+  assert(llvm::sys::fs::exists(unitFilePath));
```

Comment:
This might work in the LIT tests because there we are compiling from the `cd /` root. I would like to check this when everything is compiled not from the root.

Reply:
Just checked: C++17 has `#include <filesystem>` and this API:

```cpp
    namespace fs = std::filesystem;
    ...
    fs::path dir ("/tmp");
    fs::path file ("foo.txt");
    fs::path full_path = dir / file;
    std::cout << full_path << std::endl;
```

But on my macOS 10.14 is says this is unavailable until 10.15.

Reply:
You can re-use this part. I suppose: https://github.com/mull-project/mull/blob/master/lib/SourceLocation.cpp#L52

re: `std::filesystem` so far I don't see a reason to use it, we can just rely on the LLVM's fs/path modules and call it a day :)